Performance: Use dynamic imports

miina opened the pull request:
## Context
This PR does some code splitting by dynamically importing modules not required when loading the editor at first.

Note: there are still more modules to import dynamically, such as `draft-js` and some others.

## To-do

- [x] html-to-image
- [x] react-color (needed for the color picker only)
- draft-js (seemingly needed for text elements only) [EDIT: WIll be done in a follow-up issue: google/web-stories-wp/issues/8336]
- [x] colorthief (could be loaded only when starting to use, that's happening in the background anyway)
- [x] react-calendar (used for the calendar only)
- [x] react-moveable (not needed for the initial render)
- [x] Fix tests

## Testing Instructions

### QA

This PR can be tested by following these steps:

I
1. Add a shape
2. Now set network to slow 3g
3. Open the Color Picker
4. Verify that while the color picker is loading, a spinner is displayed and once it loads, the color picker is not positioned partially out of the screen.

II
1. Open the Document panel.
2. Set the network to slow 3g
3. Open the calendar.
4. Verify that while the calendar is loading, a spinner is displayed, and calendar doesn't "jump" once it loads.


## Checklist

<!-- Check these after PR creation -->
- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8251

miina committed:
Dynamic import for html-to-image.

miina committed:
Dynamic import for react-calendar.

swissspidy commented on assets/src/edit-story/components/moveable/index.js:
```diff
@@ -27,13 +26,22 @@ import InOverlay from '../overlay';
 
+const Moveable = lazy(() =>
+  import(
+    /* webpackPrefetch: true */ /* webpackChunkName: "chunk-react-moveable" */ 'react-moveable'
```
Nit: this could be combined to `/* webpackPrefetch: true, webpackChunkName: "chunk-react-moveable" */`

miina committed:
Adjust tests.

miina committed:
More test updates.

miina committed:
Nit + lints.

miina committed:
Dynamic import for colorthief.

miina committed:
Dynamic import for react-color.

miina committed:
react-color dynamic import.

swissspidy commented on assets/src/edit-story/components/form/color/colorInput.js:
```diff
@@ -121,6 +121,9 @@ const StyledSwatch = styled(Swatch)`
   ${focusStyle};
 `;
 
+const loadReactColor = () =>
```
All these dynamic imports might need a `.catch()` in case webpack fails to load them for whatever reason, not sure.

miina replied:
It looks like using `ErrorBoundary` seems to be the recommended way for catching issues with it. We already have those per area, not sure if it would make sense to add more. Thoughts?

miina committed:
Fix some tests.

miina committed:
Fix some  warnings.

miina committed:
Fix more warnings.

miina committed:
Lint

miina committed:
Adjust tests.

miina committed:
Adjust test some more.

barklund commented on assets/src/edit-story/components/colorPicker/currentColorPicker.js:
```diff
@@ -17,14 +17,35 @@
 import { __ } from '@web-stories-wp/i18n';
 import { useFeature } from 'flagged';
 
+const Saturation = lazy(() =>
+  import(
+    /* webpackChunkName: "chunk-react-color" */ 'react-color/lib/components/common'
+  ).then((module) => ({ default: module.Saturation }))
+);
+const Hue = lazy(() =>
+  import(
+    /* webpackChunkName: "chunk-react-color" */ 'react-color/lib/components/common'
+  ).then((module) => ({ default: module.Hue }))
+);
+const Alpha = lazy(() =>
+  import(
+    /* webpackChunkName: "chunk-react-color" */ 'react-color/lib/components/common'
+  ).then((module) => ({ default: module.Alpha }))
+);
+
```
Just want to be sure this doesn't actually include the same module 3 times (as in tripling the source code) for weird reasons, right?

miina replied:
Yep, importing from the same file should include it just once.

barklund commented on assets/src/edit-story/utils/getMediaBaseColor.js:
```diff
@@ -65,7 +62,12 @@ function getDefaultOnloadCallback(nodeKey, resolve, reject) {
       const node = document.body[nodeKey];
-      resolve(thief.getColor(node.firstElementChild));
+      import(
+        /* webpackPrefetch: true, webpackChunkName: "chunk-colorthief" */ 'colorthief'
```
Any idea why the file for this one is *not* called `vendors-chunk-colorthief` like the other files, but just `chunk-colorthief`? @swissspidy?

swissspidy replied:
This is due to the size. The default threshold is 3KB for a separate chunk to be generated for loading separately.

Docs: https://v4.webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks

barklund commented on assets/src/edit-story/utils/getMediaBaseColor.js:
```diff
@@ -65,7 +62,12 @@ function getDefaultOnloadCallback(nodeKey, resolve, reject) {
+        resolve(thief.getColor(node.firstElementChild));
+      });
     } catch (err) {
       trackError('image_base_color', err.message);
```
Does this catch actually catch anything now or is it superfluous?

miina replied:
Hm, we can add `catch` directly to the import now, instead.

barklund commented on assets/src/edit-story/components/colorPicker/currentColorPicker.js:
```diff
@@ -151,49 +172,51 @@ function CurrentColorPicker({ rgb, hsl, hsv, hex, onChange, showOpacity }) {
-        {showOpacity && (
-          <AlphaWrapper>
-            <Alpha
+        <Suspense fallback={null}>
```
I think we want a fallback of the correct size in here, as otherwise the popup is placed before the container has the right size (and thus dips below the viewport) - and it also somehow gets a horizontal scrollbar?

Note that I have the connection speed throttled to "Slow 3G" to force this.

![slow-react-color](https://user-images.githubusercontent.com/637548/125837537-fb24aca0-73ea-4ff0-84df-c85788e1b7f7.gif)

miina replied:
Added:
![Screenshot 2021-07-16 at 11 34 59](https://user-images.githubusercontent.com/3294597/125918577-bfad1a44-2663-4f5e-a5e0-c24d3ee2da6b.png)

barklund commented on assets/src/edit-story/components/form/dateTime/datePicker.js:
```diff
@@ -88,38 +91,46 @@ function DatePicker({ currentDate, onChange, onViewChange }) {
-        )}
-      />
-    </CalendarWrapper>
+    <Suspense fallback={null}>
```
Here I also think a properly sized fallback would make the swap a bit cleaner. Perhaps even with a spinner? (again note the throttling)

![slow-calendar](https://user-images.githubusercontent.com/637548/125837866-37d29c77-8917-496a-804f-7ffbf3648dcb.gif)

miina replied:
Added:
![Screenshot 2021-07-16 at 11 37 25](https://user-images.githubusercontent.com/3294597/125918825-494ade7e-a6a7-4962-a249-a0c519e1d1bd.png)

barklund reviewed (commented):
Great job here - everything seems to work perfectly and I'm not noticing even the slightest lag anywhere on a regular connection.

I also tested the components with a slow connection and the calendar and color picker could probably do with a properly-sized fallback to reduce the visual blink - and perhaps a simple spinner too.

miina committed:
Remove try-catch.

miina committed:
Add spinner to the color picker.

miina committed:
Add fallback for calendar.

merapi commented on assets/src/edit-story/components/panels/document/publish/publishTime.js:
```diff
@@ -117,6 +120,7 @@ function PublishTime() {
           ref={dateFieldRef}
           activeItemLabel={displayLabel}
           selectButtonStylesOverride={focusStyle}
+          onPointerEnter={() => loadCalendar()}
```
Do we also want `onFocus={() => loadCalendar()}` here?

miina replied:
Yep, good point!

merapi reviewed (approved):
One nit. Always nice to see improvement in resource loading/processing time 🚀

miina committed:
Load calendar when focusing, too

Commit:
Merge branch 'main' into perf/8251-dynamic-imports

Commit:
Merge branch 'perf/8251-dynamic-imports' of github.com:google/web-stories-wp into perf/8251-dynamic-imports

swissspidy committed:
Remove some incorrect assertions from e2e tests

`[data-testid="FrameElement"]` does not exist. It's `frameElement` with a lowercase `f`.

So these assertions would have never failed.

Also, there is always at least 1 `[data-testid="frameElement"]` element: the page background.

swissspidy committed:
Improve cleanup in tests

swissspidy committed:
Fix e2e tests (?)

swissspidy committed:
Und part of the last change

swissspidy committed:
Lint fix

spacedmonkey committed:
Merge branch 'main' into perf/8251-dynamic-imports

spacedmonkey committed:
Merge branch 'main' into perf/8251-dynamic-imports

spacedmonkey committed:
Merge branch 'main' into perf/8251-dynamic-imports

Commit:
Merge branch 'main' into perf/8251-dynamic-imports

Commit:
Fixed failing test
